docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,227 @@
+to store senders and their nonces, allowing for simple multi-dimensional
+prioritization (2-ary).
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into 
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into mempool. This is to
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+> The `PriorityMempool[T MempoolTx]` implementation will support Options such as
+> limiting the mempool size by a fixed number of bytes.
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. In order to check validity of the proposed
+block, we must iterate over the list of transactions and execute them using the
+same mode/execution strategy as `CheckTx`. In other words, we execute each transaction
+using the AnteHandler only -- no messages are executed. However, we cannot just
+execute `CheckTx` again, because `BaseApp` already has a modified `checkState` at
+this point. So when executing `ProcessProposal`, we create a similar branched
```

Comment:
I lack the context of BaseApp logic to fully understand this paragraph. However, the word "similar" in this line is raising some alarms to me. Let me explain.
ALL blocks produced by `PrepareProposal` must be accepted by `ProcessProposal`, 100% of them, not 1 less. Otherwise we cannot guarantee that the chain won't halt. And I'm not talking about determinism here, but liveness. I am more than willing to give an example (in a synchronous conversation) of how such a situation would be way more painful to deal with than a panic due to AppHash mismatch (something we are all familiar with).

Back to what is explained in the paragraph, having `PrepareProposal` rely on `CheckTx` having been executed previously for the transactions Reaped seems _a priori_ correct to me . BUT, the checks done by `ProcessProposal` MUST return EXACTLY the same outcome for each transaction as was returned by `CheckTx`.

Reply:
There is a case where validators may vote to reject process proposal. This is one of the key points of process proposal. If the proposer constructs a block with an invalid tx, the other validators have the right to reject the block.  Seems like this is not the assumption made by tendermint? 

While a low chance there is a chance it could happen.

Reply:
Yes, but as in original Tendermint, correct validators never reject a block from a correct proposer. If the block is created according to the code, it will pass the validity checks by correct validators. If a Byzantine proposer proposes an invalid block, the validators can reject it. The difference between original Tendermint and ABCI++ is that, in ABCI++, validity is extended to the application-specific content of a block (the transactions).

Thus, in ABCI++, if a proposer proposes a block with a transaction that a correct validator deems invalid, there are two cases:
1) the proposer is Byzantine
2) the proposer follows the code, but the code has a bug (either at PrepareProposal or ProcessProposal).

Reply:
I'm a bit confused here in terms of what we want to do in `ProcessProposal` (@prettymuchbryce this might be of relevance to you and your team).

> ALL blocks produced by PrepareProposal must be accepted by ProcessProposal, 100% of them, not 1 less.

So what's the point of `ProcessProposal` even running validation? What happens when it fails or returns an error? Does the chain halt? What should we do here?

Reply:
If ProcessProposal returns an error, the validator does not prevote for that block. Thus Tendermint Consensus will go to the next round, and a new proposer can construct a new block (for the same height). Considering the two cases from above:

- If we are in Case 1. eventually we will have a round with a correct proposer, and we will decide on its block and make progress. All good.
- If we are in Case 2 (there is a bug), we might keep on increasing round numbers without ever deciding on a block for that height. (From the outside it looks like the chain halted)

Reply:
At the risk of spoiling @josef-widder's clear and concise explanation, let me add the following:
* Think of Josef's Case 1 as some _rogue_ operator that hacks its implementation of `PrepareProposal`/`ProcessProposal` (so, it deviates from released SDK or App code) for some obscure purpose. Tendermint makes the assumption (it always has) that the rogue operator(s) won't ever reach 1/3 of the total voting power. So, `ProcessProposal` can reject proposals from the rogue validators and the chain won't halt  (as Josef explained above).
* In Case 2, there is a bug in **released** code, so very likely _all_ validators will hit it (and their combined voting power is way bigger than 1/3 of the total). Here, the risk of halting the chain is very real.

Reply:
okay this is how we understand it. I think the first comment was confusing as it does not always hold true.

Reply:
Ok, your comment made me pretty confused TBH @sergio-mena. You made it seem like returning an error should never happen and thus the chain would halt.

From what I gather, it's okay to return an error -- Tendermint will just bump the round and try again.

Reply:
OK. Josef's comments should be enough. If you understand those, then let's take those and ignore mine.

> You made it seem like returning an error should never happen and thus the chain would halt.

But that's what I meant 😀. Actually, the chain _might_ halt. It's a possibility, not a certainty. If you are still confused, let's not make this thread any longer: let's have a 20-min-long synchronous discussion where hopefully all the confusion can be cleared up.